src/Language/Docker/Parser.hs

```diff
@@ -34,19 +32,40 @@ dockerfile =
     return $ InstructionPos i (T.pack . sourceName $ pos) (unPos . sourceLine $ pos)
 
 parseText :: Text -> Either Error Dockerfile
-parseText = parse (contents dockerfile) "<string>" . dos2unix
+parseText txt = do
+  let ?esc = findEscapePragma (T.lines (dos2unix txt))
+   in parse (contents dockerfile) "<string>" (dos2unix txt)
 
 parseFile :: FilePath -> IO (Either Error Dockerfile)
-parseFile file = doParse <$> B.readFile file
-  where
-    doParse = parse (contents dockerfile) file . dos2unix . E.decodeUtf8With E.lenientDecode
+parseFile file = doParse file <$> B.readFile file
 
 -- | Reads the standard input until the end and parses the contents as a Dockerfile
 parseStdin :: IO (Either Error Dockerfile)
-parseStdin = doParse <$> B.getContents
+parseStdin = doParse "/dev/stdin" <$> B.getContents
+
+-- | Parses a list of lines from a dockerfile one by one until either the escape
+-- | pragma has been found, or pragmas are no longer expected.
+-- | Pragmas can occur only until a comment, an empty line or another
+-- | instruction occurs (i.e. they have to be the first lines of a Dockerfile).
+findEscapePragma :: [Text] -> Char
+findEscapePragma [] = defaultEsc
+findEscapePragma (l:ls) =
+  case parse (contents parseComment) "<line>" l of
+    Left _ -> defaultEsc
+    Right (Pragma (Escape (EscapeChar c))) -> c
+    Right (Pragma _) -> findEscapePragma ls
+    Right _ -> defaultEsc
   where
-    doParse = parse (contents dockerfile) "/dev/stdin" . dos2unix . E.decodeUtf8With E.lenientDecode
+    ?esc = defaultEsc
+
+doParse :: FilePath -> B.ByteString -> Either Error Dockerfile
+doParse path txt =
+  let ?esc = findEscapePragma (T.lines (dos2unix (E.decodeUtf8With E.lenientDecode txt)))
+   in parse (contents dockerfile) path (dos2unix (E.decodeUtf8With E.lenientDecode txt))
```

Comment:
Can we save the work of decoding the file twice and move that into another variable?

Reply:
Sure. That makes sense.